{% extends "layout.html" %}
{% block loads %}
    <!-- jQuery and jQuery UI -->
    <link rel="stylesheet" type="text/css" href="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/themes/smoothness/jquery-ui.css" />
    <script type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jqueryui/1.11.2/jquery-ui.min.js"></script>

    <!-- Google Fonts and Font awesome -->
    <link rel="stylesheet" type="text/css" href='//fonts.googleapis.com/css?family=PT+Sans:400,700' />
    <link rel="stylesheet" type="text/css" href='//fonts.googleapis.com/css?family=Open+Sans' />
    <!-- ALREADY LOADED BY layout.html: link rel="stylesheet" type="text/css" href='//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css' / -->
{% endblock %}
{% block body %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.analyse_variants = {{ analyse_variants|tojson|safe }};

        function getLinks(){
            console.log("test");
            return "hallo";
        }


    </script>

    <script>
        $(document).ready(function() {

            window.variants_analyse = _.template($('#analyse-table-template').html());
            $('#analyse_table_container').html(variants_analyse({analyse_variants: analyse_variants}));

            $('#frequency_table').tablesorter({
                stringTo: 'bottom',
                sortList: [[4,1], [0,0]],
                headers: {
                    4: {
                        sorter: false
                    }
                }
            });


            function get_af_category(d) {
                var allele_count = d.allele_count;
                var allele_freq =  d.allele_freq;

                if (!allele_freq) {
                    return [0, '0'];
                } else if (allele_freq < 1/10000) {
                    return [2, '<1/10000'];
                } else if (allele_freq < 1/1000) {
                    return [3, '1/10000-0.001'];
                } else if (allele_freq < 1/100) {
                    return [4, '0.001-0.01'];
                } else if (allele_freq < 1/20) {
                    return [5, '0.01-0.05'];
                } else if (allele_freq < 1/2) {
                    return [6, '0.05-0.5'];
                } else {
                    return [7, '0.5-1'];
                }
            }

            var data = window.analyse_variants;



            var width = 50;
            var height = 15;

            var x_scale = d3.scale.linear()
                    .domain([0, 7])
                    .range([0, width]);

            var svg;
            $.each(data, function(i, d) {
                d3.select('#analyse_af_box_' + d.variant_id).attr("data-tooltip", "Shows allele frequency \n on a discrete " +
                        "scale: \n singletons, <1/10,000, \n <1/1000, <1%, <5%, \n <50%, >50%. \n This particular variant is \n " +
                        get_af_category(d)[1] + ".");
                svg = d3.select('#analyse_af_box_' + d.variant_id)
                        .append("svg")
                        .attr("width", width)
                        .attr("height", height)
                        .append("g");

                for (var j=0; j<8; j++) {
                    svg.append('rect')
                            .style('stroke', 'steelblue')
                            .style('fill', 'white')
                            .attr('x', x_scale(j))
                            .attr('y', 0)
                            .attr('height', height)
                            .attr('width', x_scale(1) - x_scale(0))
                }

                svg.append('rect')
                        .style('fill', 'steelblue')
                        .attr('x', 0)
                        .attr('y', 0)
                        .attr('width', function() {
                            return x_scale(get_af_category(d)[0]);
                        })
                        .attr('height', height);

            });
        });

    </script>

    <div id="analyse_table_container"></div>

    <script type="text/template" id="analyse-table-template">
    <div class="row">
        <div class="col-md-12">
            <div id="frequency_info_container">
                <div class="section_header">Analyse Overview</div>
                <div id="frequency_table_container" class="frequency_displays">
                    <table id="frequency_table">
                        <thead>
                        <tr>
                            <th class="tooltip-table-header omit_csv" data-tooltip="chr:pos reference/alternate &#xa;(dbSNP135 rsID)">
                                Variant
                            </th>
                            <th>Chrom</th>
                            <th>Position</th>
                            <th class="hidden-xs tooltip-table-header"
                                data-tooltip="HGVS annotation (protein change, where defined, or transcript change for splice variants, otherwise empty)">
                                Consequence
                            </th>
                            <th class="hidden-xs tooltip-table-header" data-tooltip="VQSR sensitivity filter & hard filters">Filter</th>
                            <th class="tooltip-table-header"
                                data-tooltip="Variant Effect Predictor (VEP) &#xa; annotation using Gencode 81.&#xa; Worst across all transcripts of this gene.">
                                Annotation
                            </th>
                            <th class="tooltip-table-header"
                                data-tooltip="Various flags (LoF from LOFTEE, MNP) &#xa; that may change annotation/confidence">Flags
                            </th>
                            <th class="tooltip-table-header"
                                data-tooltip="Alternate allele count in genotypes &#xa;(genotype quality >=20 & depth >=10)">Allele
                                Count
                            </th>
                            <th class="hidden-xs tooltip-table-header"
                                data-tooltip="Total number of called genotypes &#xa;(genotype quality >= 20 & depth >=10)">Allele Number
                            </th>
                                <th class="hidden-xs tooltip-table-header"
                                    data-tooltip="Number of homozygous individuals for this alternate allele">Number of Homozygotes
                                </th>
                                <th class="hidden-xs tooltip-table-header"
                                    data-tooltip="Number of hemizygous individuals for this alternate allele">Number of Hemizygotes
                                </th>
                            <th class="tooltip-table-header allele_frequency_column"
                                data-tooltip="Frequency using only &#xa; high-quality genotypes &#xa; The graphic displays &#xa; allele frequency on a  &#xa; discrete scale: Singletons, &#xa; <1/10,000, <1/1000, <1%, &#xa; <5%, <50%, >50%">
                                Allele Frequency
                            </th>
                            <th>Disease Genes</th>
                            <th class='hidden-xs'>HRC</th>
                            <th class='hidden-xs'>ESP</th>
                            <th class='hidden-xs'>ExAC</th>
                            <th class='hidden-xs'>Kaviar</th>
                            <th class='hidden-xs'>1000Genomes</th>
                            <th class='hidden-xs'>DANN</th>
                            <th class='hidden-xs'>FATHMM</th>
                            <th class='hidden-xs'>CADD</th>
                            <th class='hidden-xs'>dbscSNV</th>
                            <th class='hidden-xs'>interpro-domain</th>
                            <th class='hidden-xs'>gerp</th>
                            <th class='hidden-xs'>phylo_placental</th>
                            <th class='hidden-xs'>phylo_vertebrate</th>
                            <th class='hidden-xs'>Annotation</th>
                        </tr>
                        </thead>
                        <tbody>
                        <% _.each(analyse_variants, function(variant) { %>
                        <% if (variant.filter == "PASS") { %>
                        <tr>
                                <td class="omit_csv">
                                    <a href="/variant/<%= variant.chrom %>-<%= variant.pos %>-<%= variant.ref %>-<%= variant.alt %>"
                                       target="_blank">
                                        <%= variant.chrom %>:<%= variant.pos %> <br>
                                        <% if (variant.ref.length > 20) { %>
                                        <%= variant.ref.slice(0,20) %>...
                                        <% } else { %>
                                        <%= variant.ref %>
                                        <% } %> /
                                        <% if (variant.alt.length > 20) { %>
                                        <%= variant.alt.slice(0,20) %>...
                                        <% } else { %>
                                        <%= variant.alt %>
                                        <% } %>
                                        <% if (variant.rsid != '.') { %>
                                        <br>
                                        (<%= variant.rsid %>)
                                        <% } %>
                                    </a>
                                </td>
                                <td class='hidden-xs'> <%= variant.chrom %></td>
                                <td class='hidden-xs'> <%= variant.pos %></td>
                            <td class='hidden-xs'>
                                <% if (variant.HGVS != '') { %>
                                <% if (variant.HGVS.length > 20) { %>
                                <%= variant.HGVS.slice(0,20) %>...<% } else { %>
                                <%= variant.HGVS %><% } %>
                                <% } %>
                            </td>
                            <td class='hidden-xs'> <%= variant.filter %></td>
                            <td class="<%= variant.category %>" data-consequence="<%= variant.major_consequence %>"><b>
                                <% if (variant.major_consequence) { %>
                                <%= variant.major_consequence.replace('_variant', '').replace(/_/g, ' ').replace('utr',
                                'UTR').replace('3 prime', "3'").replace('5 prime', "5'").replace('nc ', "non-coding ") %>
                                <% } %>
                            </b></td>
                            <td>
                                <% _.each(variant.flags, function(flag) { %>
                                <% if (flag == 'LoF flag') { span_tag = 'warning' } else { span_tag = 'danger' } %>
                                <a href="/variant/<%= variant.chrom %>-<%= variant.pos %>-<%= variant.ref %>-<%= variant.alt %>"
                                   target="_blank"><span class="label label-<%= span_tag %> tooltip-table-header"
                                                         data-tooltip="Variant annotation dubious: &#xa; see variant page for details"><%= flag %></span></a>
                                <% }); %>
                            </td>
                            <td>
                                <%= variant.allele_count %>
                            </td>
                            <td class='hidden-xs'>
                                <%= variant.allele_num %>
                            </td>
                            <td class="hidden-xs"> <%= variant.hom_count %> </td>
                            <td class="hidden-xs">
                                <%= variant.hemi_count %>
                            </td>
                            <td>
                                <%= parseFreq(variant.allele_freq) %>
                                <div id="analyse_af_box_<%= variant.variant_id %>" style="float: right;"></div>
                            </td>
                            <td class="omit_csv">
                                <% _.each(variant.disease_genes, function(gene) { %>
                                <a href="/gene/<%= gene.ensembl %>"
                                   target="_blank">
                                    <%= gene.uniprot %>
                                </a>
                                <% }); %>
                            </td>
                            <td class='hidden-xs'>
                                <%= parseFreq(variant.HRC_AF) %>
                            </td>
                            <td class='hidden-xs'>
                                <%= parseFreq(variant.ESP) %>
                            </td>
                            <td class='hidden-xs'>
                                <%= parseFreq(variant.ExAC_AF) %>
                            </td>
                            <td class='hidden-xs'>
                                <%= parseFreq(variant.Kaviar_AF) %>
                            </td>
                            <td class='hidden-xs'>
                                <%= parseFreq(variant.g1k) %>
                            </td>
                            <td class='hidden-xs'>
                                <% if (variant.DANN) { %>
                                <%= Number(variant.DANN).toPrecision(3) %>
                                <% } else { %>
                                <%= variant.DANN %>
                                <% } %>
                            </td>
                            <td class='hidden-xs'>
                                <% if (variant.FATHMM) { %>
                                <%= Number(variant.FATHMM).toPrecision(3) %>
                                <% } else { %>
                                <%= variant.FATHMM %>
                                <% } %>
                            </td>
                            <td class='hidden-xs'>
                                <% if (variant.CADD) { %>
                                <%= Number(variant.CADD).toPrecision(3) %>
                                <% } else { %>
                                <%= variant.CADD %>
                                <% } %>
                            </td>
                            <td class='hidden-xs'><%= variant.dbscSNV_RF_SCORE %></td>
                            <td class='hidden-xs'><%= variant.interpro_domain %></td>
                            <td class='hidden-xs'><%= variant.GERP %></td>
                            <td class='hidden-xs'><%= variant.phylo_placental %></td>
                            <td class='hidden-xs'><%= variant.pyhlo_vertebrate %></td>
                            <td class='hidden-xs'><%= variant.annotation %></td>
                        </tr>
                        <% } %>
                        <% }); %>
                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    </div>



    </script>

{% endblock %}
