<script type="text/javascript">
    window.is_gene = typeof window.transcripts_in_gene != 'undefined';
</script>
<script type="text/template" id="variant-table-template">
<span>
    <a class="btn btn-success hidden-xs" id="export_to_csv">Export table to CSV</a>
</span>
{% if transcripts_in_gene is defined %}
    <div>
        <small>&dagger; denotes a consequence that is for a non-canonical transcript</small>
    </div>
{% endif %}
<table id="variant_table" class="tablesorter">
    <thead>
        <tr>
            <th class="tooltip-table-header omit_csv" data-tooltip="chr:pos reference/alternate &#xa;(dbSNP135 rsID)">
                Variant</th>
            <th class='hidden-xs'>Chrom</th>
            <th class='hidden-xs'>Position</th>
            <td class="hidden">RSID</td>
            <th class="hidden">Reference</th>
            <th class="hidden">Alternate</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="HGVS annotation (protein change, where defined, or transcript change for splice variants, otherwise empty)">Consequence</th>
            <th class="hidden">Protein Consequence</th>
            <th class="hidden">Transcript Consequence</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="VQSR sensitivity filter & hard filters">Filter</th>
            <th class="tooltip-table-header" data-tooltip="Variant Effect Predictor (VEP) &#xa; annotation using Gencode 81.&#xa; Worst across all transcripts of this gene.">Annotation</th>
            <th class="tooltip-table-header" data-tooltip="Various flags (LoF from LOFTEE, MNP) &#xa; that may change annotation/confidence">Flags</th>
            <th class="tooltip-table-header" data-tooltip="Alternate allele count in genotypes &#xa;(genotype quality >=20 & depth >=10)">Allele Count</th>
            <th class="hidden-xs tooltip-table-header" data-tooltip="Total number of called genotypes &#xa;(genotype quality >= 20 & depth >=10)">Allele Number</th>
            {% if chrom != 'Y' %}
                <th class="hidden-xs tooltip-table-header" data-tooltip="Number of homozygous individuals for this alternate allele">Number of Homozygotes</th>
            {% endif %}
            {% if chrom == 'X' or chrom == 'Y' %}
                <th class="hidden-xs tooltip-table-header" data-tooltip="Number of hemizygous individuals for this alternate allele">Number of Hemizygotes</th>
            {% endif %}
            <th class="tooltip-table-header allele_frequency_column" data-tooltip="Frequency using only &#xa; high-quality genotypes &#xa; The graphic displays &#xa; allele frequency on a  &#xa; discrete scale: Singletons, &#xa; <1/10,000, <1/1000, <1%, &#xa; <5%, <50%, >50%">
                Allele Frequency
            </th>
            <th class='hidden-xs'>HRC</th>
            <th class='hidden-xs'>ESP</th>
            <th class='hidden-xs'>ExAC</th>
            <th class='hidden-xs'>Kaviar</th>
            <th class='hidden-xs'>1000Genomes</th>
            <th class='hidden-xs'>CLINSIG</th>
            <th class='hidden-xs'>CLNDBN</th>
            <th class='hidden-xs'>CLNACC</th>
            <th class='hidden-xs'>DANN</th>
            <th class='hidden-xs'>FATHMM</th>
            <th class='hidden-xs'>CADD</th>
            <th class='hidden-xs'>dbscSNV</th>
            <th class='hidden-xs'>interpro-domain</th>
            <th class='hidden-xs'>gerp</th>
            <th class='hidden-xs'>phylo_placental</th>
            <th class='hidden-xs'>phylo_vertebrate</th>
        </tr>
    </thead>
    <tbody>
    <% _.each(table_variants, function(variant) { %>
        <tr class="table_variant" id="variant_<%= variant.variant_id %>"
            category="<%= variant.category %>"
            filter_status="<%= variant.filter %>"
            frequency="<%= variant.allele_freq %>"
            major_consequence="<%= variant.major_consequence %>"
            indel="<%= variant.indel %>"
          >
            <td class="omit_csv">
                <a href="/variant/<%= variant.chrom %>-<%= variant.pos %>-<%= variant.ref %>-<%= variant.alt %>" target="_blank">
                <%= variant.chrom %>:<%= variant.pos %>
                <% if (variant.ref.length > 20) { %>
                    <%= variant.ref.slice(0,20) %>...
                <% } else { %>
                    <%= variant.ref %>
                <% } %> /
                <% if (variant.alt.length > 20) { %>
                    <%= variant.alt.slice(0,20) %>...
                <% } else { %>
                    <%= variant.alt %>
                <% } %>
                <% if (variant.rsid != '.') { %>
                    (<%= variant.rsid %>)
                <% } %>
                </a>
            </td>
            <td class='hidden-xs'> <%= variant.chrom %></td>
            <td class='hidden-xs'> <%= variant.pos %></td>
            <td class="hidden"><%= variant.rsid %></td>
            <th class="hidden"> <%= variant.ref %></th>
            <th class="hidden"> <%= variant.alt %></th>
            <td class='hidden-xs'>
                <% if (variant.HGVS != '') { %>
                    <% if (variant.HGVS.length > 20) { %>
                        <%= variant.HGVS.slice(0,20) %>...<% } else { %>
                        <%= variant.HGVS %><% } %><% if (is_gene & variant.CANONICAL != 'YES') { %><span class="tooltip-table-header" data-tooltip="Annotation is for non-canonical transcript">&dagger;</span><% } %>
                <% } %>
            </td>
            <td class='hidden'> <%= variant.HGVSp %></td>
            <td class='hidden'> <%= variant.HGVSc %></td>
            <td class='hidden-xs'> <%= variant.filter %> </td>
            <td class="<%= variant.category %>" data-consequence="<%= variant.major_consequence %>"><b>
                <% if (variant.major_consequence) { %>
                    <%= variant.major_consequence.replace('_variant', '').replace(/_/g, ' ').replace('utr', 'UTR').replace('3 prime', "3'").replace('5 prime', "5'").replace('nc ', "non-coding ") %>
                <% } %>
            </b> </td>
            <td>
                <% _.each(variant.flags, function(flag) { %>
                    <% if (flag == 'LoF flag') { span_tag = 'warning' } else { span_tag = 'danger' } %>
                    <a href="/variant/<%= variant.chrom %>-<%= variant.pos %>-<%= variant.ref %>-<%= variant.alt %>" target="_blank"><span class="label label-<%= span_tag %> tooltip-table-header" data-tooltip="Variant annotation dubious: &#xa; see variant page for details"><%= flag %></span></a>
                <% }); %>
            </td>

            <td>
                <% if(window.selected_population  == "All") { %>
                <%= variant.allele_count %>
                <% } else { %>
                <%  var i; %>
                <%for (i in variant.cohorts) { %>
                <% if(variant.cohorts[i].name  == window.selected_population) { %>
                <%= variant.cohorts[i].allele_count %>
                <% } %>
                <%}%>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if(window.selected_population  == "All") { %>
                <%= variant.allele_num %>
                <% } else { %>
                <%  var i; %>
                <%for (i in variant.cohorts) { %>
                <% if(variant.cohorts[i].name  == window.selected_population) { %>
                <%= variant.cohorts[i].allele_num %>
                <% } %>
                <%}%>
                <% } %>
            </td>
            {% if chrom != 'Y' %}
                <td class='hidden-xs'>
                    <% if(window.selected_population  == "All") { %>
                    <%= variant.hom_count %>
                    <% } else { %>
                    <%  var i; %>
                        <%for (i in variant.cohorts) { %>
                            <% if(variant.cohorts[i].name  == window.selected_population) { %>
                                <%= variant.cohorts[i].hom_count %>
                             <% } %>
                        <%}%>
                    <% } %>
                </td>
                <!--<td class="hidden-xs"> <%= variant.hom_count %> </td>-->
            {% endif %}
            {% if chrom == 'X' or chrom == 'Y' %}
                <td class="hidden-xs"> <%= variant.hemi_count %> </td>
            {% endif %}
            <td>
                <% if(window.selected_population  == "All") { %>
                <% if (variant.allele_freq) { %>
                <%= variant.allele_freq.toPrecision(4) %>
                <% } else { %>
                <%= variant.allele_freq %>
                <% } %>
                <% } else { %>
                <%  var i; %>
                <%for (i in variant.cohorts) { %>
                <% if(variant.cohorts[i].name  == window.selected_population) { %>
                <% if (variant.cohorts[i].allele_freq) { %>
                <%= variant.cohorts[i].allele_freq.toPrecision(4) %>
                <% } else { %>
                <%= variant.cohorts[i].allele_freq %>
                <% } %>
                <% } %>
                <%}%>
                <% } %>
                <div id="variant_af_box_<%= variant.variant_id %>" style="float: right;"></div>
            </td>
            <td class='hidden-xs'>
                <% if (variant.HRC_AF) { %>
                <%= Number(variant.HRC_AF).toPrecision(3) %>
                <% } else { %>
                <%= variant.HRC_AF %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.ESP) { %>
                <%= Number(variant.ESP).toPrecision(3) %>
                <% } else { %>
                <%= variant.ESP %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.ExAC_MAF) { %>
                <%= Number(variant.ExAC_MAF).toPrecision(3) %>
                <% } else { %>
                <%= variant.ExAC_MAF %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.Kaviar_AF) { %>
                <%= Number(variant.Kaviar_AF).toPrecision(3) %>
                <% } else { %>
                <%= variant.Kaviar_AF %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.g1k) { %>
                <%= Number(variant.g1k).toPrecision(3) %>
                <% } else { %>
                <%= variant.g1k %>
                <% } %>
            </td>
            <td class='hidden-xs'><%= variant.CLINSIG.replace("|", "\n")  %></td>
            <td class='hidden-xs'><%= variant.CLNDBN.replace(",", "\n") %></td>
            <td class='hidden-xs'><%= variant.CLNACC.replace("|", "\n") %></td>
            <td class='hidden-xs'>
                <% if (variant.DANN) { %>
                <%= Number(variant.DANN).toPrecision(3) %>
                <% } else { %>
                <%= variant.DANN %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.FATHMM) { %>
                <%= Number(variant.FATHMM).toPrecision(3) %>
                <% } else { %>
                <%= variant.FATHMM %>
                <% } %>
            </td>
            <td class='hidden-xs'>
                <% if (variant.CADD) { %>
                <%= Number(variant.CADD).toPrecision(3) %>
                <% } else { %>
                <%= variant.CADD %>
                <% } %>
            </td>
            <td class='hidden-xs'><%= variant.dbscSNV_ADA_SCORE %></td>
            <td class='hidden-xs'><%= variant.interpro_domain %></td>
            <td class='hidden-xs'><%= variant.GERP %></td>
            <td class='hidden-xs'><%= variant.phylo_placental %></td>
            <td class='hidden-xs'><%= variant.pyhlo_vertebrate %></td>
        </tr>
    <% }); %>
    </tbody>
</table>
</script>